### DISCUSSION TITLE
Contribution guidelines - squashed commits reason

### DISCUSSION THREAD

**Discussion by tdeo** [rubocop/rubocop#11100] [Ideas]
Hello,

I was curious to hear a bit more context about the reason behind requiring squashed commits in Pull requests in the contribution guidelines. 

I believe keeping the commit history in a pull requests adds some clarity in regards to implementation steps and the evolutions throughout reviews, and also helps reviewer see only the difference since their last review (for that to work properly, you also want to merge master into your branch along the way instead of rebasing).

**Comment by koic** [comment]
This is my opinion as a long time active maintainer.

What I want as a maintainer is atomic reason for change, not implementation steps. I'm basically not interested in the implementation steps and intermediate work in pull requests. Most of the time I re-review the entire PR. And, GitHub has "Compare" button when I need it, and that's enough.

IMHO, the great thing about `git rebase` is that developer can emulate write clean code the first time, regardless of developer individual steps. For example, whether it's test-first, test-last is irrelevant to the reason for the change. Great growth as an SCM that CVS and Subversion couldn't do.
Furthermore, the rails/rails repository, which has been maintained for many years, has a similar operation.

I explained something similar recently.

- https://github.com/rubocop/rubocop/pull/10839#issuecomment-1225074841
- https://github.com/rubocop/rubocop-rails/pull/747#issuecomment-1190094697

To summarize, the clean history of the trunk commits is important in maintenance. Thank you.

**Comment by tdeo** [comment]
Thanks for sharing the links, I see the "maintaining the commit hash" as a valid reason and it's indeed not the case when using "Squash and merge".

I was thinking about that because of the specifity of https://github.com/rubocop/rubocop/pull/11058, in which I believe exhibiting multiple commits to show that the random nature of the test has been eliminated and I believe truely assists the reviewer. Of course I'll squash and rebase the PR once it has been reviewed.

**Reply by bbatsov** [reply]
I think about this as making sure that within a PR there are just a many commits as needed and all internal steps (e.g. for test purposes) are removed for the sake of keeping the commit history clean - sometimes this is just one commit at the end, sometimes it's several. I use the "squash and commit" only when I feel a PR maps cleanly to a single commit. In other cases I'd ask the contributors to squash their commits manually in whatever way makes the most sense for the particular PR.

**Comment by dvandersluis** [comment]
I like to treat commits the same way we would treat classes as per the single responsibility principle. IMO, each commit should be a discrete unit where all the changes belong together. For example, if I was going to refactor something and then add new functionality, I'd do the refactor as one commit and the new behaviour as a second commit.

When it comes to PR review or merging, commits should be squashed together into these units. Oftentimes that results in a single commit, but like @bbatsov said, it should be just as many as needed, with WIP commits squashed.

**Comment by pirj** [comment]
Just 4c from me.

1. Squashing is for **related** commits. This gives an idea about topics, what was fixed/improved and why. Keeping checkpoint commits goes against that, as checkpoint commits can be intermingled addressing different topics. "Related" appears [in the pull request template](https://github.com/rubocop/rubocop/blame/cf0dd9d06d6ceb1d67c9fd167953bee4c54b85cf/.github/PULL_REQUEST_TEMPLATE.md#L12)

2. Squashing is done right before merging the PR. I find it troublesome to deal with frequent squashing, as it requires to `git branch -D <branch-name>` before `hub pr <pr-number>` when the PR is squashed and force-pushed. Might be a local bug, or just my lack of understanding, and there might be a way around this.

**Comment by koic** [comment]
This is a different perspective on refactoring work and new feature additions against version-control commits. The Refactoring Book "Refactoring: Improving the Design of Existing Code (2nd Edition)" page 53 has the following mentions, which are worth remembering:

> Too often, the refactorings are closely interwoven with adding new features, and it's not worth the time to separate them out. This can also remove the context for refactoring, making the refactoring commits hard to justify.

https://martinfowler.com/books/refactoring.html

I like this pragmatic logic as a maintainer.

**Comment by tdeo** [comment]
Hello, 
Just wanted to add an additional detail on that topic - I've been using the code from [this PR of rubocop-rspec](https://github.com/rubocop/rubocop-rspec/pull/1377) to be able to write custom rubocop-rspec cops with specs, and I must say in that context, having the pull request branches being force-pushed (because of rebasing / squashing) frequently causes our gem installation to break because the commit reference disappears.

**Reply by bquorning** [reply]
Ah yes, we could have merged that PR a bit sooner. I believe we will soon release a v2.15.0 which should provide a more stable gem reference 😅

If you find yourself in a similar situation again in the future, you could always make a fork of the repository and install from that. That would probably be a more stable target.

**Comment by pirj** [comment]
Our apologies for this inconvenience. @Darhazer @bquorning do you have any objections on merging this PR?

**Reply by Darhazer** [reply]
Go ahead and merge!

**Reply by tdeo** [reply]
Thanks !
